### ISSUE TITLE
Parser edge case allows the use of arbitrary expressions as targets

### ISSUE THREAD

**Issue by asleep-cult** [microsoft/pyright#2589]
**Describe the bug**
The parser allows you to use any (non-target) expression as a target.

**To Reproduce**
1. Create a file
2. Write some code that uses a non-target expression as a target e.g. `0 if 0 else 0 = 0`
3. Run Pyright on it
4. Observe how nothing happens

**Expected behavior**
Pyright should report a parse error.

**Screenshots or Code**
```python
x() = 0
0 = 0
'' = 0
0 if 0 else 0 = 0
1 + 2 = 0
1 == 2 = 0
{} = 0

for '' in range(10):
    pass

with x() as 0:
    pass
```

**VS Code extension or command-line**
Command line pyright 1.1.188 

**Additional context**
I initially found this bug while reading through this function:
https://github.com/microsoft/pyright/blob/main/packages/pyright-internal/src/parser/parser.ts#L3876

**Comment by erictraut** [comment]
The parser implementation should catch and report these errors. I'm not able to repro the problem you're seeing.

Here's the output I receive when I run pyright on the sample code above:

```
  test.py:1:1 - error: Expression cannot be assignment target (reportGeneralTypeIssues)
  test.py:2:1 - error: Expression cannot be assignment target (reportGeneralTypeIssues)
  test.py:3:1 - error: Expression cannot be assignment target (reportGeneralTypeIssues)
  test.py:4:1 - error: Expression cannot be assignment target (reportGeneralTypeIssues)
  test.py:5:1 - error: Expression cannot be assignment target (reportGeneralTypeIssues)
  test.py:6:1 - error: Expression cannot be assignment target (reportGeneralTypeIssues)
  test.py:7:1 - error: Expression cannot be assignment target (reportGeneralTypeIssues)
  test.py:9:5 - error: Expression cannot be assignment target (reportGeneralTypeIssues)
  test.py:12:6 - error: "x" is not defined (reportUndefinedVariable)
  test.py:12:13 - error: Expression cannot be assignment target (reportGeneralTypeIssues)
```

Perhaps you have suppressed errors for this file? Do you have a pyrightconfig.json file? If so, does it include an `ignore` configuration?

**Comment by asleep-cult** [comment]
Oh wow! I was convinced that this was a bug after looking through the parser implementation and getting no errors from the CLI or Pylance. But to answer your question, no I do not have any file called `pyrightconfig.json`. After trying it in a different directory I am seeing the errors in the CLI but I only get the errors from Pylance when turning on basic type checking. Have I managed to screw up the configuration or is this behavior expected?

**Comment by erictraut** [comment]
Ah, I see what's going on. These errors are not caught within the parser. These statements are permitted by the Python grammar, so the parser allows them. They are instead caught during the semantic analysis phase, which is why they're reported using the `reportGeneralTypeIssues` rule. I think it's reasonable to report these unconditionally though, since they're not really related to typing. I've made that change, which will appear in the next release. With this change, you won't need to enable `reportGeneralTypeIssues` (which is part of the "basic" type checking mode) to receive these errors.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.189, which I just published. It will also be included in the next release of pylance.
